packages/model/addon/-private/system/promise-many-array.ts

```diff
@@ -0,0 +1,322 @@
+import { assert } from '@ember/debug';
+import { dependentKeyCompat } from '@ember/object/compat';
+import { tracked } from '@glimmer/tracking';
+import Ember from 'ember';
+
+import { resolve } from 'rsvp';
+
+import { DEPRECATE_EVENTED_API_USAGE } from '@ember-data/private-build-infra/deprecations';
+
+/**
+ @module @ember-data/model
+ */
+/**
+  This class is returned as the result of accessing an async hasMany relationship
+  on an instance of a Model extending from `@ember-data/model`.
+
+  A PromiseManyArray is an array-like proxy that also proxies certain method calls
+  to the underlying ManyArray in addition to being "promisified".
+   
+  Right now we proxy:
```

Comment:
We now proxy more methods right?

Reply:
This comment was pulled forward from the existing implementation, and describes only the methods on ManyArray that aren't typical to MutableArray implementations. We of course also proxy all MutableArray methods just by the nature of being an ArrayProxy.